// postprocessing/FilmPass.js
import {
	ShaderMaterial,
	UniformsUtils
} from '../../three.module.js';
import { Pass, FullScreenQuad } from './Pass.js';
import { FilmShader } from '../shaders/FilmShader.js';

class FilmPass extends Pass {

	constructor( noiseIntensity, scanlinesIntensity, scanlinesCount, grayscale ) {
		
		super();

		if ( FilmShader === undefined ) console.error( 'THREE.FilmPass relies on FilmShader' );

		const shader = FilmShader;

		this.uniforms = UniformsUtils.clone( shader.uniforms );

		this.material = new ShaderMaterial( {

			uniforms: this.uniforms,
			vertexShader: shader.vertexShader,
			fragmentShader: shader.fragmentShader

		} );

		if ( grayscale !== undefined )	this.uniforms.grayscale.value = grayscale;
		if ( noiseIntensity !== undefined ) this.uniforms.nIntensity.value = noiseIntensity;
		if ( scanlinesIntensity !== undefined ) this.uniforms.sIntensity.value = scanlinesIntensity;
		if ( scanlinesCount !== undefined ) this.uniforms.sCount.value = scanlinesCount;

		this.fsQuad = new FullScreenQuad( this.material );

	}

	render( renderer, writeBuffer, readBuffer, deltaTime /*, maskActive */ ) {

		this.uniforms[ 'tDiffuse' ].value = readBuffer.texture;
		this.uniforms[ 'time' ].value += deltaTime;

		if ( this.renderToScreen ) {

			renderer.setRenderTarget( null );
			this.fsQuad.render( renderer );

		} else {

			renderer.setRenderTarget( writeBuffer );
			if ( this.clear ) renderer.clear();
			this.fsQuad.render( renderer );

		}

	}

	get getNoiseIntensity() {
		/* get構文 */
		return this.noiseIntensity;
	}
	get getScanlinesIntensity() {
		/* get構文 */
		return this.scanlinesIntensity;
	}
	get getScanlinesCount() {
		/* get構文 */
		return this.scanlinesCount;
	}
	
	get getGrayscale() {
		/* get構文 */
		return this.grayscale;
	}
	
	set setNoiseIntensity(s) {
        /* set構文 */
        this.uniforms.nIntensity.value = s
    }
	set setScanlinesIntensity(s) {
        /* set構文 */
        this.uniforms.sIntensity.value = s
    }
	set setScanlinesCount(s) {
        /* set構文 */
        this.uniforms.sCount.value = s
    }
	set setGrayscale (s) {
        /* set構文 */
        this.uniforms.grayscale.value = s
    }

}

export { FilmPass };
